### PR TITLE
Added serialized size in the message header

### PR DESCRIPTION
**Pull request by nvidianz**
### Description
1. Added serialized size in message header as PAYLOAD_LEN so caller can get the size.
2. Added format_size function to format size into human-readable format like (MB, MiB etc)
3. Reformatted the getTask message to display correct size.

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Review by yhwen — DISMISSED** [review]
LGTM

File: nvflare/fuel/f3/cellnet/utils.py [hunk 1]
```diff
@@ -93,3 +99,38 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
...
+    else:
+        kilo = 1000.0
+        suffix = "B"
+
+    unit_found = False
+    for unit in ("", "K", "M", "G", "T", "P", "E", "Z"):
+        if abs(num) < kilo:
+            unit_found = True
+            break
+        num /= 1000.0
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/fuel/f3/cellnet/utils.py:127]: suggest adding unit test cases specifically to this method

File: nvflare/fuel/f3/cellnet/utils.py [hunk 2]
```diff
@@ -79,8 +79,14 @@ def encode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
             message.payload = fobs.dumps(message.payload, buffer_list=True)
         message.set_header(encoding_key, encoding)
 
+    size = len(message.payload) if message.payload else 0
+    message.set_header(MessageHeaderKey.PAYLOAD_LEN, size)
+
 
 def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCODING):
+    size = len(message.payload) if message.payload else 0
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py:87]: The header should be there in most cases but we have to deal with the special case that 2 different versions of server/client. This is very cheap so we just do it twice.

File: nvflare/fuel/f3/cellnet/utils.py [hunk 3]
```diff
@@ -93,3 +99,38 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
...
+    Args:
+        num: Size in bytes
+        binary: If binary, one K is 1024 bytes, otherwise 1000 bytes.
+
+    Returns: Size in human-readable format (like 10MB, 100.2GiB etc)
+
+    """
+
+    if binary:
+        kilo = 1024.0
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py:116]: So we can get fraction of kilo in the output. For example, 1.5KB

File: nvflare/fuel/f3/cellnet/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py]: Good catch! Added test cases.

File: nvflare/fuel/f3/cellnet/utils.py [hunk 4]
```diff
@@ -93,3 +99,38 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
...
+
+    unit_found = False
+    for unit in ("", "K", "M", "G", "T"):
+        if abs(num) < kilo:
+            unit_found = True
+            break
+        num /= kilo
+
+    if not unit_found:
+        unit = "P"
```
Comment (YuanTingHsieh) [inline nvflare/fuel/f3/cellnet/utils.py:130]: A question regarding implementation, why "P" needs to be outside of the for loop?

File: nvflare/fuel/f3/cellnet/utils.py [hunk 5]
```diff
@@ -93,3 +99,38 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
         # assume to be bytes
         pass
     message.remove_header(encoding_key)
+
+
+def format_size(num, binary=False):
```
Comment (YuanTingHsieh) [inline nvflare/fuel/f3/cellnet/utils.py:104]: What if the "num" passed in is NOT a integer?
Should we check for that?

File: tests/unit_test/fuel/f3/cellnet/utils_test.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+        assert format_size(1000000000, False) == "1GB"
+        assert format_size(1000000000, True) == "953.7MiB"
+        assert format_size(1073741824, False) == "1.1GB"
+        assert format_size(1073741824, True) == "1GiB"
+
+        assert format_size(1000000000000, False) == "1TB"
+        assert format_size(1000000000000, True) == "931.3GiB"
+        assert format_size(1099511627776, False) == "1.1TB"
+        assert format_size(1099511627776, True) == "1TiB"
```
Comment (YuanTingHsieh) [inline tests/unit_test/fuel/f3/cellnet/utils_test.py:41]: Suggest add one case of "PB" and negative cases.

File: nvflare/fuel/f3/cellnet/utils.py (hunk 4 above)
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py:130]: The loop can only go so far, we choose 'T' as last one in the loop. P is used for any larger numbers. So there must be a line to handle this. I added a test case to demonstrate this.

File: tests/unit_test/fuel/f3/cellnet/utils_test.py (hunk 6 above)
Comment (nvidianz) [reply tests/unit_test/fuel/f3/cellnet/utils_test.py:41]: Added PB, negative, string test cases.

File: nvflare/fuel/f3/cellnet/utils.py (hunk 5 above)
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py]: Added support for numbers in string.

File: nvflare/private/fed/client/communicator.py [hunk 7]
```diff
@@ -266,8 +267,9 @@ def submit_update(
         )
         end_time = time.time()
         return_code = result.get_header(MessageHeaderKey.RETURN_CODE)
+        size = task_message.get_header(MessageHeaderKey.PAYLOAD_LEN)
```
Comment (yhwen) [inline nvflare/private/fed/client/communicator.py:270]: Here the size calculation is not correct. It always got 2. It looks like the header not been set correctly. Could you please check?

**Review by yhwen — CHANGES_REQUESTED** [review]

File: nvflare/fuel/f3/cellnet/utils.py [hunk 8]
```diff
@@ -93,3 +99,39 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
         # assume to be bytes
         pass
     message.remove_header(encoding_key)
+
+
+def format_size(size, binary=False):
```
Comment (yanchengnv) [inline nvflare/fuel/f3/cellnet/utils.py:104]: I thought "binary" should be True by default. 
But I wonder whether we need this option. What's the use case?

File: nvflare/private/fed/client/communicator.py [hunk 9]
```diff
@@ -194,14 +195,14 @@ def pull_task(self, servers, project_name, token, ssid, fl_ctx: FLContext):
         return_code = task.get_header(MessageHeaderKey.RETURN_CODE)
 
         if return_code == ReturnCode.OK:
-            size = len(task.payload)
+            size = task.get_header(MessageHeaderKey.PAYLOAD_LEN)
             task.payload = task.payload
```
Comment (yanchengnv) [inline nvflare/private/fed/client/communicator.py:199]: Please remove this line.

File: nvflare/private/fed/client/communicator.py (hunk 7 above)
Comment (yanchengnv) [reply nvflare/private/fed/client/communicator.py:270]: I think there is an error in the calculation - see my comment in core_cell.py.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
The msg size calculation seems incorrect.

File: nvflare/fuel/f3/cellnet/utils.py [hunk 10]
```diff
@@ -93,3 +99,39 @@ def decode_payload(message: Message, encoding_key=MessageHeaderKey.PAYLOAD_ENCOD
...
+        size: Size in bytes
+        binary: If binary, one K is 1024 bytes, otherwise 1000 bytes.
+
+    Returns: Size in human-readable format (like 10MB, 100.2GiB etc)
+
+    """
+
+    if binary:
+        kilo = 1024.0
+        suffix = "iB"
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py:132]: For binary Kilo (1024), the standard suffix is MiB, GiB etc.

File: nvflare/fuel/f3/cellnet/utils.py (hunk 8 above)
Comment (nvidianz) [reply nvflare/fuel/f3/cellnet/utils.py:119]: To conform with convention and standard, when we say MB, GB, it's non binary mode. That's what we have been using so left it as default.

File: nvflare/private/fed/client/communicator.py (hunk 7 above)
Comment (nvidianz) [reply nvflare/private/fed/client/communicator.py:270]: Fixed.

File: nvflare/private/fed/client/communicator.py (hunk 9 above)
Comment (nvidianz) [reply nvflare/private/fed/client/communicator.py:199]: Done

File: nvflare/private/fed/client/communicator.py [hunk 11]
```diff
@@ -194,14 +195,13 @@ def pull_task(self, servers, project_name, token, ssid, fl_ctx: FLContext):
...
-            size = len(task.payload)
-            task.payload = task.payload
+            size = task.get_header(MessageHeaderKey.PAYLOAD_LEN)
             task_name = task.payload.get_header(ServerCommandKey.TASK_NAME)
             fl_ctx.set_prop(FLContextKey.SSID, ssid, sticky=False)
             if task_name not in [SpecialTaskName.END_RUN, SpecialTaskName.TRY_AGAIN]:
                 self.logger.info(
-                    f"Received from {project_name} server "
-                    f" ({size} Bytes). getTask: {task_name} time: {end_time - start_time} seconds"
+                    f"Received from {project_name} server. getTask: {task_name} size:{format_size(size)} "
```
Comment (yanchengnv) [inline nvflare/private/fed/client/communicator.py:203]: Please add a space btw "size:" and "{format_size(size)".

**Review by yanchengnv — CHANGES_REQUESTED** [review]
LGTM, except for a small change needed in the log msg - see my comment.

File: nvflare/private/fed/client/communicator.py (hunk 11 above)
Comment (nvidianz) [reply nvflare/private/fed/client/communicator.py]: Done.

**Review by yhwen — APPROVED** [review]

**Review by yanchengnv — APPROVED** [review]
LGTM

**Pull request merged**